sdks/java/core/src/test/java/org/apache/beam/sdk/util/ReshuffleTest.java

```diff
@@ -88,6 +98,58 @@ public void testJustReshuffle() {
     pipeline.run();
   }
 
+  /**
+   * Tests that timestamps are not moved forwards in time after applying a {@link Reshuffle} with
+   * the default {@link WindowingStrategy}.
+   */
+  @Test
+  @Category(RunnableOnService.class)
+  public void testReshuffleDoesNotMoveTimestampsForwards() {
+    PCollection<KV<String, TimestampedValue<String>>> input =
+        pipeline
+            .apply(
+                Create.timestamped(
+                        TimestampedValue.of("foo", BoundedWindow.TIMESTAMP_MIN_VALUE),
+                        TimestampedValue.of("foo", new Instant(0)),
+                        TimestampedValue.of("bar", new Instant(33)),
+                        TimestampedValue.of("bar", GlobalWindow.INSTANCE.maxTimestamp()))
+                    .withCoder(StringUtf8Coder.of()))
+            .apply(
+                WithKeys.of(
+                    new SerializableFunction<String, String>() {
+                      @Override
+                      public String apply(String input) {
+                        return input;
+                      }
+                    }))
+            .apply(ReifyTimestamps.<String, String>inValues());
+
+    PCollection<TimestampedValue<TimestampedValue<String>>> output =
```

Comment:
Add a comment explaining this nested TimestampedValue structure. For a second I thought we got a monad going here.